lxd/daemon_images.go

```diff
@@ -14,9 +14,44 @@ import (
 	log "gopkg.in/inconshreveable/log15.v2"
 )
 
+type Progress struct {
+	io.Reader
+	total  int64
+	length int64
+	op     *operation
+}
+
+func (pt *Progress) Read(p []byte) (int, error) {
+	n, err := pt.Reader.Read(p)
+	if n > 0 {
+		pt.total += int64(n)
+		percentage := float64(pt.total) / float64(pt.length) * float64(100)
+
+		if pt.op != nil {
+			meta := pt.op.metadata
+			if meta == nil {
+				meta = make(map[string]interface{})
+			}
+
+			progressInt := 10 - (int(percentage) % 10) + int(percentage)
```

Comment:
Rounding to the closest multiple of 10 limits the number of notifications we sent out. Otherwise we'd be updating the operation and send log messages and event notifications for every single byte we download.